### PR TITLE
[PM-14818] Adjust instructions for running EF integration tests

### PR DESCRIPTION
**Pull request by addisonbeck**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-10863

## 📔 Objective

This commit adds up to date documentation for running entity framework integration tests for `server`. It has a companion PR for adding some of the mentioned functionality in https://github.com/bitwarden/server/pull/4912

## 📸 Screenshots

Screenshot 2024-10-17 at 7 54 35 PM

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by rkac-bw — DISMISSED** [review]
lgtm

File: docs/getting-started/server/database/ef/index.mdx [hunk 1]
```diff
@@ -196,30 +196,48 @@ If you would like to verify that everything worked correctly:
...
+  "databases:0:enabled": "true",
+  "databases:1:type": "Sqlite",
+  "databases:1:enabled": "true",
+  "databases:1:connectionString": "Data Source=_________",
+  "databases:2:type": "MySql",
+  "databases:2:connectionString": "server=localhost;uid=root;pwd=_________;database=ef_test",
+  "databases:2:enabled": "true",
+  "databases:3:type": "SqlServer",
+  "databases:3:connectionString": "Server=localhost;Database=ef_test;User Id=SA;Password=_________;Encrypt=True;TrustServerCertificate=True;",
+  "databases:3:enabled": "true"
```
Comment (eliykat) [inline docs/getting-started/server/database/ef/index.mdx:238]: I suggested [changing the format](https://github.com/bitwarden/server/pull/4912#discussion_r1811719851) of these user secrets. If you do that, this will need to be updated. If you don't want to do that, then we should at least note here (maybe in a callout) that the index of each type must be maintained for the migrate script to work.

File: docs/getting-started/server/database/ef/index.mdx [hunk 2]
```diff
@@ -196,30 +196,48 @@ If you would like to verify that everything worked correctly:
...
+  "postgreSql": {
+    "connectionString": "Host=localhost;Username=postgres;Password=_________;Database=vault_dev"
+  },
+  "mySql": {
+    "connectionString": "server=localhost;uid=root;pwd=_________;database=vault_dev"
+  },
+  "sqlite": {
+    "connectionString": "Data Source=/path/to/bitwardenServer/repository/server/dev/db/bitwarden.sqlite"
+  },
+  ```
```
Comment (eliykat) [inline docs/getting-started/server/database/ef/index.mdx:220]: I don't think this is required. It's already covered in the sections above so this is largely repeating information. I think you could just start with the following paragraph and it would work just as well (or better because it's not repeating information).

File: docs/getting-started/server/database/ef/index.mdx [hunk 3]
```diff
@@ -196,30 +196,48 @@ If you would like to verify that everything worked correctly:
...
-dotnet user-secrets set "Ef:Postgres" "[POSTGRES_CONNECTION_STRING]"
-dotnet user-secrets set "Ef:MySql" "[MYSQL_CONNECTION_STRING]"
-
-# You can also set the connection string for your normal development MS SQL database like below
-dotnet user-secrets set "Dapper:SqlServer" "[MSSQL_CONNECTION_STRING]"
-```
-
-You can then run just those tests from the `test/Infrastructure.IntegrationTest` folder using
-`dotnet test`.
+In your `server/dev/secrets.json` file find these two blocks of secrets:
```
Comment (mzieniukbw) [inline docs/getting-started/server/database/ef/index.mdx:199]: ❓ The`server/` prefix might be confusing for some. This name is not guaranteed and we never reference it anywhere. It's almost always `secrets.json`.

File: docs/getting-started/server/database/ef/index.mdx [hunk 4]
```diff
@@ -196,30 +196,48 @@ If you would like to verify that everything worked correctly:
...
+  "databases:2:type": "MySql",
+  "databases:2:connectionString": "server=localhost;uid=root;pwd=_________;database=ef_test",
+  "databases:2:enabled": "true",
+  "databases:3:type": "SqlServer",
+  "databases:3:connectionString": "Server=localhost;Database=ef_test;User Id=SA;Password=_________;Encrypt=True;TrustServerCertificate=True;",
+  "databases:3:enabled": "true"
+  ```
+
+With connection strings applied to your projects: ensure your databases are all migrated using
+`pwsh server/dev/migrate.ps1 --all`. Then you can run EF tests from the
```
Comment (mzieniukbw) [inline docs/getting-started/server/database/ef/index.mdx:242]: ❓ The`server/` prefix might be confusing for some and it's not guaranteed.

File: docs/getting-started/server/database/ef/index.mdx [hunk 5]
```diff
@@ -196,30 +196,48 @@ If you would like to verify that everything worked correctly:
...
-# https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-6.0&tabs=windows#secret-manager
-dotnet user-secrets set "Ef:Postgres" "[POSTGRES_CONNECTION_STRING]"
-dotnet user-secrets set "Ef:MySql" "[MYSQL_CONNECTION_STRING]"
-
-# You can also set the connection string for your normal development MS SQL database like below
-dotnet user-secrets set "Dapper:SqlServer" "[MSSQL_CONNECTION_STRING]"
-```
-
-You can then run just those tests from the `test/Infrastructure.IntegrationTest` folder using
-`dotnet test`.
```
Comment (mzieniukbw) [inline docs/getting-started/server/database/ef/index.mdx:225]: I like this text more.

File: docs/getting-started/server/database/ef/index.mdx (hunk 3 above)
Comment (addisonbeck) [reply docs/getting-started/server/database/ef/index.mdx:199]: I think including the repo name in paths in a set of documentation that references many repositories is helpful. I'd like to keep this. While not guaranteed, I suppose, the repository name is the default folder name when you clone a repository.

**Comment by addisonbeck** [comment]
I've merged in the changes from https://github.com/bitwarden/contributing-docs/pull/480 so we can merge these as a complete package with links from one to the other.

File: docs/getting-started/server/database/ef/index.mdx (hunk 4 above)
Comment (addisonbeck) [reply docs/getting-started/server/database/ef/index.mdx:231]: Responded above

File: docs/getting-started/server/database/ef/index.mdx (hunk 2 above)
Comment (addisonbeck) [reply docs/getting-started/server/database/ef/index.mdx]: Good idea. Removed on latest.

File: docs/getting-started/server/database/ef/index.mdx (hunk 1 above)
Comment (addisonbeck) [reply docs/getting-started/server/database/ef/index.mdx]: I think we're going to end up keeping the format, so I have not addressed this

File: docs/getting-started/server/database/ef/index.mdx [hunk 6]
```diff
@@ -196,30 +196,37 @@ If you would like to verify that everything worked correctly:
...
+:::
 
-# You can also set the connection string for your normal development MS SQL database like below
-dotnet user-secrets set "Dapper:SqlServer" "[MSSQL_CONNECTION_STRING]"
-```
+This block is used for test databases for each supported provider type. These are what integration
+tests will connect to. You should update the password for these connection strings to match your
+existing databases if you have not already. If these settings are not present at all in your
+`secrets.json` file just add them to the bottom. These settings _do not_ go in `globalSettings`.
+Then run `pwsh setup_secrets.ps1 -clear` to apply them to your local projects.
```
Comment (mzieniukbw) [inline docs/getting-started/server/database/ef/index.mdx:228]: For consistency, let's add `server/dev/` path prefix

**Review by mzieniukbw — CHANGES_REQUESTED** [review]

File: docs/getting-started/server/database/ef/index.mdx [hunk 7]
```diff
@@ -196,30 +196,37 @@ If you would like to verify that everything worked correctly:
...
-dotnet user-secrets set "Ef:MySql" "[MYSQL_CONNECTION_STRING]"
+:::
 
-# You can also set the connection string for your normal development MS SQL database like below
-dotnet user-secrets set "Dapper:SqlServer" "[MSSQL_CONNECTION_STRING]"
-```
+This block is used for test databases for each supported provider type. These are what integration
+tests will connect to. You should update the password for these connection strings to match your
+existing databases if you have not already. If these settings are not present at all in your
+`secrets.json` file just add them to the bottom. These settings _do not_ go in `globalSettings`.
```
Comment (addisonbeck) [inline docs/getting-started/server/database/ef/index.mdx:227]:
```suggestion
`server/dev/secrets.json` file just add them to the bottom. These settings _do not_ go in `globalSettings`.
```

**Review by mzieniukbw — DISMISSED** [review]

**Review by rkac-bw — DISMISSED** [review]

File: docs/contributing/testing/database/index.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
+# Database Integration Testing
+
+Since Bitwarden has multiple database options, testing them all automatically is incredibly
+important so that we don't have to have a full QA process for each database type. This is where the
+`Infrastructure.IntegrationTest` project comes in, allowing the setup of tests similarly to how the
+databases are consumed in the applications through their common interfaces, generally an
+`I*Repository`. These tests are automatically executed on all supported databases in our testing
+pipeline.
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:8]: I find this a bit hard to follow and is probably packing too much info into the opening paragraph. How about:

```suggestion
Since Bitwarden has multiple database options, automated testing is extremely important to avoid time-consuming manual testing for each database type. The `Infrastructure.IntegrationTest` project allows you to write tests that are automatically executed on all supported databases in our testing pipeline.
```

File: docs/contributing/testing/database/index.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+## Creating a new test
+
+To create a new database test just add the `[DatabaseTheory]` and `[DatabaseData]` attributes to
+test. Now, in the parameters of the test you are able to "inject" any repository layer services
+directly into the test. The test will run for every database that is
+[configured in the current environment](#configure-the-tests). Since you inject the interface of the
+service, some runs of the test will use the Dapper based repository implementation targeting
+Microsoft SQL Server and others will use the Entity Framework Core based implementations, which we
+use for MySql, Postgres, and SQLite.
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:18]:
```suggestion
To create a new database test, add the `[DatabaseTheory]` and `[DatabaseData]` attributes to
test. Then, use the parameters of the test to inject any repository layer services
you need. The test will run for every database that is
[configured in the current environment](#configure-the-tests). Since you inject the interface of the
service, some runs of the test will use the Dapper-based repository implementation targeting
Microsoft SQL Server and others will use the Entity Framework Core based implementations (which we
use for MySql, Postgres, and SQLite).
```

File: docs/contributing/testing/database/index.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+The goal of database tests is to test the business logic that is encapsulated in a given method. If
+the stored procedure in SQL Server then calls another procedure to update the
+`User.AccountRevisionDate` then the same EF implementation should do that as well. By running the
+test against all variants we are ensuring all the variants are feature-equal. Locally, you may want
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:23]:
```suggestion
The goal of database tests is to test the business logic that is encapsulated in a given method. For example, if
a stored procedure in SQL Server calls another procedure to update the
`User.AccountRevisionDate` then the corresponding EF implementation should do that as well. By running the
test against all variants, we are ensuring all the variants are feature-equal.
```

File: docs/contributing/testing/database/index.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+to only run the SQL Server tests along with one EF implementation; SQLite is often the easiest in
+that situation. This may work well for a very long time and save you some time overall but there are
+differences between the EF database providers such that you will one day get errors in the CI
+pipeline.
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:27]: I'm not sure what the doc is recommending here. Should we use SQLite because it's easy? Or all 3 unsupported-supported EF databases to be comprehensive?

Given that it runs in the pipeline I think we can recommend the easier path for local development.

```suggestion
to only run the SQL Server tests along with one EF implementation; SQLite is often the easiest in
that regard. The other supported EF database providers will still run in the pipeline to catch any
differences between them.
```

File: docs/contributing/testing/database/index.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+## Configure the databases
+
+The databases are expected to have the latest migrations applied.
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:31]: Link to your instructions on this?

File: docs/contributing/testing/database/index.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+## Configure the tests
+
+The tests are configured through
+[.NET Configuration](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration) in the
+order they are applied: user secrets, environment variables prefixed with `BW_TEST_`, and command
+line args.
```
Comment (eliykat) [inline docs/contributing/testing/database/index.md:38]:
```suggestion
The tests are configured through
[.NET Configuration](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration). They are applied in the following order: user secrets, environment variables prefixed with `BW_TEST_`, and command
line args.
```

File: docs/contributing/testing/database/index.md (hunk 12 above)
Comment (addisonbeck) [reply docs/contributing/testing/database/index.md]: They are already linked below for running the tests locally. I don't really think this section is offering anything and opted to remove it instead.

**Review by eliykat — APPROVED** [review]
